File: tools/schedule.py
"""Tool: class_schedule — jadwal batch mendatang per kelas."""
from db import fetchall

VALID_CLASSES = {"N5", "N4", "N3", "JFT", "KAIWA"}


def class_schedule(class_name: str) -> dict:
    """
    Ambil jadwal batch mendatang beserta harga dan kuota untuk satu kelas.

    Args:
        class_name: Nama kelas. Nilai valid: N5 | N4 | N3 | JFT | KAIWA

    Returns:
        dict dengan keys: schedule (list batch), quota (dict batch_name -> sisa)
    """
    cls = class_name.strip().upper()
    if cls not in VALID_CLASSES:
        return {"error": f"class_name harus salah satu dari {sorted(VALID_CLASSES)}"}

    sql = """
        SELECT id, batch_name, batch_desc, batch_date, batch_price,
               batch_quota, batch_status
        FROM batch_registrations
        WHERE batch_date >= CURDATE()
          AND batch_desc LIKE %s
          {kaiwa_exclude}
        ORDER BY batch_date ASC
    """
    like_cls = f"%{cls}%"
    if cls != "KAIWA":
        sql = sql.format(kaiwa_exclude="AND batch_desc NOT LIKE %s")
        rows = fetchall(sql, (like_cls, f"%kaiwa{cls}%"))
    else:
        sql = sql.format(kaiwa_exclude="")
        rows = fetchall(sql, (like_cls,))

    # group quota by batch_name
    quota: dict[str, int] = {}
    for r in rows:
        name = r["batch_name"]
        quota[name] = quota.get(name, 0) + (r["batch_quota"] or 0)

    return {
        "schedule": rows,
        "quota": quota,
    }

File: tools/schedule.js
import { fetchAll } from '../db.js';

const VALID_CLASSES = new Set(['N5', 'N4', 'N3', 'JFT', 'KAIWA']);

export async function classSchedule({ class_name }) {
  const cls = class_name.trim().toUpperCase();
  if (!VALID_CLASSES.has(cls)) {
    return { error: `class_name harus salah satu dari: ${[...VALID_CLASSES].join(', ')}` };
  }

  const likeParam = `%${cls}%`;
  let rows;

  if (cls !== 'KAIWA') {
    rows = await fetchAll(
      `SELECT id, batch_name, batch_desc, batch_date, batch_price, batch_quota, batch_status
       FROM batch_registrations
       WHERE batch_date >= CURDATE()
         AND batch_desc LIKE ?
         AND batch_desc NOT LIKE ?
       ORDER BY batch_date ASC`,
      [likeParam, `%kaiwa${cls}%`],
    );
  } else {
    rows = await fetchAll(
      `SELECT id, batch_name, batch_desc, batch_date, batch_price, batch_quota, batch_status
       FROM batch_registrations
       WHERE batch_date >= CURDATE()
         AND batch_desc LIKE ?
       ORDER BY batch_date ASC`,
      [likeParam],
    );
  }

  const quota = {};
  for (const r of rows) {
    quota[r.batch_name] = (quota[r.batch_name] ?? 0) + (r.batch_quota ?? 0);
  }

  return { schedule: rows, quota };
}
